### PR TITLE
Json diff for release.

### PR DESCRIPTION
**Pull request by kevingranade**
#### Summary
None

#### Purpose of change
Generate and apply a json diff for the release.
Added a script and instructions for how to generate said diff.
Also generated a diffstat and estimate of number of contributors.
It's an estimate because a count of unique author strings overcounts by a lot, so I went through and deduplicated where I could see repeated names or email addresses.

#### Describe the solution
The jq script filters out just entities we're interested in (objects with a type and an id or ident field), then outputs blocks of ids with the associated type as a header.
The bash script then generates a diff between two such documents and filters them down to just identifiers of a given type that are only present in the later version.
By annotating this diff with line numbers and then using an AWK script, we can ennumerate the number of entities added of each type.
The final step is pretty messy and manual, I took this list of types-with-counts and manually assigned them to various groupings and cleaned up the names to be friendlier, and sorted the entries by descending order of magnitude.

#### Describe alternatives you've considered
More scripting, less scripting, not doing it, making someone else do it.
The format I ended up with here is not writ in stone, if you have an idea for better readability or better section separation, I'd like to hear it.  Tweaking category groupings is also a possibility.

#### Testing
Rerun the scripts and check my math, if you dare.

#### Additional context
There are also very likely some typos and math errors sprinkled around.

### CODE REVIEW

**Comment by actual-nh** [comment]
I need to head to work soon, but a quick comment: A `/_/ /` fix on the final result would be nice (`item groups` instead of `item_groups`, for instance). After that and anything similar is done, I can do some further copyediting (e.g., formatting, probably including more consistency in capitalization unless someone wants to modify the scripts for that).

**Comment by jbytheway** [comment]
For the last release each section heading was in all caps (see [here](https://github.com/CleverRaven/Cataclysm-DDA/blob/f17115a89ed8b08f7f13bed0a4d6fa5d09e446a7/data/changelog.txt#L446-L485)).

**Comment by kevingranade** [comment]
Will strip the underscores momentarally.
The raw contributor list was generated with:
git log 0.E..master | grep "^Author: " | sort -u

Dropping all-caps was intentional, if you think it's better that way I'm open to arguments in that vein, but I don't think maintaining continuity is compelling (and if it is I'd rather go back and edit the previous one).

**Comment by jbytheway** [comment]
I think the all caps did help make it clearer.  As it stands the text just kind of runs together.  But maybe some other formatting (like some blank lines, or maybe bullet points?) could serve the same purpose.

**Comment by actual-nh** [comment]
> I think the all caps did help make it clearer. As it stands the text just kind of runs together. But maybe some other formatting (like some blank lines, or maybe bullet points?) could serve the same purpose.

I was thinking indentation, actually.

**Comment by kevingranade** [comment]
I was thinking of indenting the lists but not the headings, let's try that.

**Comment by actual-nh** [comment]
> Will strip the underscores momentarally.
> The raw contributor list was generated with:
> git log 0.E..master | grep "^Author: " | sort -u

I get 454 total after removing various duplicates.

File: data/changelog.txt [hunk 1]
```diff
@@ -1,5 +1,70 @@
 # 0.F (2021-04-15)
 
+3974 files changed, 5440251 insertions(+), 3904330 deletions(-)
+10,463 commits
+~486 contributors
```
Comment (actual-nh) [inline data/changelog.txt:5]:
```suggestion
~454 contributors
```

File: data/changelog.txt [hunk 2]
```diff
@@ -1,5 +1,70 @@
...
+    151 comestibles, 99 tools, 53 ammunition types
+Mapgen: 778
+    424 overmap terrains, 198 palettes, 62 start locations, 41 city buildings, 25 map extras,
+    20 overmap specials, 5 overmap locations, 3 region setting
+Crafting: 562
+    223 construction groups, 201 requirements, 60 proficiencies, 50 constructions,
+    17 recipe groups, 10 recipe categories, 1 construction category
+Achievements: 363
+    115 achievements, 107 event statistics, 100 event transformations, 26 score entries, 15 conducts
+ITEM TRAITS: 325
```
Comment (actual-nh) [inline data/changelog.txt:19]:
```suggestion
Item Traits: 325
```

File: data/changelog.txt [hunk 3]
```diff
@@ -1,5 +1,70 @@
...
+    171 item groups, 97 materials, 20 ammunition types, 18 ammo effects, 13 tool qualities,
+    2 vitamins, 3 item actions, 1 fault
+Player Traits: 263
+    144 professions, 41 effect types, 28 skills, 17 mutations, 14 activity types,
+    4 skill display types, 3 behaviors, 3 movement modes, 3 scent types, 2 morale types,
+    1 butchery requirement, 1 disease type, 1 mutation types, 1 trait group
+NPCs and NPC interaction 237
```
Comment (actual-nh) [inline data/changelog.txt:26]:
```suggestion
NPCs and NPC interactions: 237
```

File: data/changelog.txt [hunk 4]
```diff
@@ -1,5 +1,70 @@
...
 
+3974 files changed, 5440251 insertions(+), 3904330 deletions(-)
+10,463 commits
+~486 contributors
+
+New game entities (core): 4510
+Items: 987
+    228 misc items, 199 books, 196 articles of clothing, 161 guns and gun related items,
+    151 comestibles, 99 tools, 53 ammunition types
+Mapgen: 778
```
Comment (actual-nh) [inline data/changelog.txt:11]:
```suggestion
Mapgen (overmap level): 778
```

File: data/changelog.txt [hunk 5]
```diff
@@ -1,5 +1,70 @@
...
+    171 item groups, 97 materials, 20 ammunition types, 18 ammo effects, 13 tool qualities,
+    2 vitamins, 3 item actions, 1 fault
+Player Traits: 263
+    144 professions, 41 effect types, 28 skills, 17 mutations, 14 activity types,
+    4 skill display types, 3 behaviors, 3 movement modes, 3 scent types, 2 morale types,
+    1 butchery requirement, 1 disease type, 1 mutation types, 1 trait group
+NPCs and NPC interaction 237
+    255 talk topics, 29 missions, 29 scenarios, 15 npcs, 6 npc classes, 3 factions
+Map Traits: 234
+    149 terrains, 12 weather types, 64 furnitures, 5 gates, 3 field types, 1 terrain transformation
```
Comment (actual-nh) [inline data/changelog.txt:29]:
```suggestion
Mapgen (detailed): 234
    149 terrains, 12 weather types, 64 furniture, 5 gates, 3 field types, 1 terrain transformation
```

File: data/changelog.txt [hunk 6]
```diff
@@ -1,5 +1,70 @@
...
+Monsters: 228
+    151 monsters, 72 harvest entries, 4 species, 1 anatomy entry
+Vehicles: 140
+    80 vehicle parts, 41 vehicles, 12 vehicle part categories, 7 vehicle groups,
```
Comment (actual-nh) [inline data/changelog.txt:33]:
```suggestion
    80 vehicle parts, 41 vehicles, 12 vehicle part categories, 7 vehicle groups
```

File: data/changelog.txt [hunk 7]
```diff
@@ -1,5 +1,70 @@
...
+Magic: 40
+    23 spells, 7 enchantments, 6 emit definitions, 4 relic definitions
+Misc: 363
+    258 json flags, 104 ascii art, 1 loot zone
+
+New game entities (mods): 3231
+Items: 773
+    179 misc items, 172 books, 161 comestibles, 85 tools, 83 articles of clothing,
+    65 guns and gun related items, 21 ammunitions, 7 compact bionic modules
```
Comment (actual-nh) [inline data/changelog.txt:42]:
```suggestion
    65 guns and gun-related items, 21 ammunition types, 7 compact bionic modules
```

File: data/changelog.txt [hunk 8]
```diff
@@ -1,5 +1,70 @@
...
+Player traits: 749
+    273 mutations, 214 trait groups, 96 professions, 64 techniques, 47 scenarios, 21 effect types,
+    9 martial arts, 8 bionics, 8 start locations, 6 mutation categories, 2 proficiencies, 1 skill
+Mapgen: 430
```
Comment (actual-nh) [inline data/changelog.txt:46]:
```suggestion
Mapgen (overmap): 430
```

File: data/changelog.txt [hunk 9]
```diff
@@ -1,5 +1,70 @@
...
+    154 item groups, 202 overmap terrains, 39 palettes, 30 overmap specials,
+    5 city buildings, 2 map extras
+Monsters: 406
+    360 monsters, 32 harvest entries, 11 species, 3 monster attacks
```
Comment (actual-nh) [inline data/changelog.txt:50]:
```suggestion
    360 monster types, 32 harvest entries, 11 species, 3 monster attacks
```

File: data/changelog.txt [hunk 10]
```diff
@@ -1,5 +1,70 @@
...
+NPCs and NPC interaction: 234
```
Comment (actual-nh) [inline data/changelog.txt:51]:
```suggestion
NPCs and NPC interactions: 234
```

File: data/changelog.txt [hunk 11]
```diff
@@ -1,5 +1,70 @@
...
+    144 talk topics, 59 npc classes, 15 mission definitions, 13 npcs, 3 factions
+Map traits: 190
+    103 terrains, 56 furnitures, 10 terrain transformations,
```
Comment (actual-nh) [inline data/changelog.txt:54]:
```suggestion
Mapgen (detailed): 190
    103 terrains, 56 furniture, 10 terrain transformations,
```

File: data/changelog.txt [hunk 12]
```diff
@@ -1,5 +1,70 @@
...
+    9 gates, 6 field types, 4 traps, 2 emit definitions
+Vehicles: 185
+    141 vehicle parts, 27 vehicle groups, 17 vehicles
+Magic: 155
+    132 spells, 23 enchantments
+Crafting: 48
+    32 requirements, 7 tool qualities, 5 construction groups, 2 constructions,
+    1 recipe category, 1 recipe group
+Item traits: 39
```
Comment (actual-nh) [inline data/changelog.txt:63]:
```suggestion
Item Traits: 39
```

File: doc/RELEASE_DIFF.md [hunk 13]
```diff
@@ -0,0 +1,24 @@
...
+jq -s -f tools/json_tools/jq/enumerate_json_entities.jq $(find data/json -name "*.json" | grep -v obsolete) > json_types_0.F
+jq -s -f tools/json_tools/jq/enumerate_json_entities.jq $(find data/mods -name "*.json" | grep -v obsolete) > json_mod_types_0.F
+
+Generate diffs between the corresponding enumerations capturing the type headers and just the added entity ids.
+diff -U 10000 json_types_0.E json_types_0.F | grep "\(^.  \"type\"\)\|\(^+ \+\".*\",\)" > json_types_diff_0.E_to_0.F
+diff -U 10000 json_mod_types_0.E json_mod_types_0.F | grep "\(^.  \"type\"\)\|\(^+ \+\".*\",\)" > json_mod_types_diff_0.E_to_0.F
+
+Run this goofy script to just list the type headers and append the count of added entities after each line.
+paste <(cut json_types_diff_0.E_to_0.F -d : -f 3) <(cut -d : -f 1 json_types_diff_0.E_to_0.F | awk 'NR == 1{old = $1; next} {print $1 - old - 1; old = $1}') > json_types_diff_0.E_to_0.F_with_counts
+paste <(cut json_mod_types_diff_0.E_to_0.F -d : -f 3) <(cut -d : -f 1 json_mod_types_diff_0.E_to_0.F | awk 'NR == 1{old = $1; next} {print $1 - old - 1; old = $1}') > json_mod_types_diff_0.E_to_0.F_with_counts
```
Comment (actual-nh) [inline doc/RELEASE_DIFF.md:24]:
```suggestion
paste <(cut json_mod_types_diff_0.E_to_0.F -d : -f 3) <(cut -d : -f 1 json_mod_types_diff_0.E_to_0.F | awk 'NR == 1{old = $1; next} {print $1 - old - 1; old = $1}') > json_mod_types_diff_0.E_to_0.F_with_counts

Then replace `_` with ` `:
```
Add however you prefer as a method - there are lots of possibilities.

**Review by actual-nh — COMMENTED** [review]
A bit of cleanup, and fixing the (approximate) number of contributors.

File: data/changelog.txt [hunk 14]
```diff
@@ -1,5 +1,70 @@
...
+Player Traits: 263
+    144 professions, 41 effect types, 28 skills, 17 mutations, 14 activity types,
+    4 skill display types, 3 behaviors, 3 movement modes, 3 scent types, 2 morale types,
+    1 butchery requirement, 1 disease type, 1 mutation types, 1 trait group
+NPCs and NPC interaction 237
+    255 talk topics, 29 missions, 29 scenarios, 15 npcs, 6 npc classes, 3 factions
+Map Traits: 234
+    149 terrains, 12 weather types, 64 furnitures, 5 gates, 3 field types, 1 terrain transformation
+Monsters: 228
+    151 monsters, 72 harvest entries, 4 species, 1 anatomy entry
```
Comment (actual-nh) [inline data/changelog.txt:31]:
```suggestion
    151 monster types, 72 harvest entries, 4 species, 1 anatomy entry
```

File: data/changelog.txt (hunk 5 above)
Comment (kevingranade) [reply data/changelog.txt:29]: It's not really part of mapgen though, it's definitions of things that appear on the map.

File: data/changelog.txt (hunk 4 above)
Comment (kevingranade) [reply data/changelog.txt:11]: These are the entities involved in determining what goes where, hence mapgen.
The map traits are the entities that define what the properties of things on the map are.

Better names are welcome, but they're definitely not both "mapgen".

File: doc/RELEASE_DIFF.md (hunk 13 above)
Comment (kevingranade) [inline doc/RELEASE_DIFF.md:24]:
```suggestion
paste <(cut json_mod_types_diff_0.E_to_0.F -d : -f 3) <(cut -d : -f 1 json_mod_types_diff_0.E_to_0.F | awk 'NR == 1{old = $1; next} {print $1 - old - 1; old = $1}') > json_mod_types_diff_0.E_to_0.F_with_counts

Then reformat into a human-readable summary.
```

File: data/changelog.txt (hunk 4 above)
Comment (actual-nh) [reply data/changelog.txt:11]: OK. I'm not the best at naming things, but... map objects? Map components? Map entities? Traits, at least to me, indicates something that would be over the entire map (or at least most of it), not a particular tile.

Comment (ToxiClay) [reply data/changelog.txt:11]: ~~It's hard for me to make any sort of judgment call without seeing the thousand-foot view.~~

Having looked at the broader context a little bit, what about "Overmap components" and "Map components"?

Comment (actual-nh) [reply data/changelog.txt:11]:
```suggestion
Overmap Components: 778
```

File: data/changelog.txt (hunk 5 above)
Comment (actual-nh) [reply data/changelog.txt:29]:
```suggestion
Map Components: 234
    149 terrains, 12 weather types, 64 furniture, 5 gates, 3 field types, 1 terrain transformation
```

File: data/changelog.txt (hunk 8 above)
Comment (actual-nh) [reply data/changelog.txt:46]:
```suggestion
Overmap Components: 430
```

File: data/changelog.txt (hunk 11 above)
Comment (actual-nh) [reply data/changelog.txt:54]:
```suggestion
Map Components: 190
    103 terrains, 56 furniture, 10 terrain transformations,
```

**Review by actual-nh — COMMENTED** [review]
How's this?

File: data/changelog.txt (hunk 4 above)
Comment (kevingranade) [reply data/changelog.txt:11]: It's not overmap components either, it's a mixture of overmap and map level entities.  The distinguishing characteristic is these are what determines what map elements go where, and they are only referenced during mapgen.

File: data/changelog.txt (hunk 5 above)
Comment (kevingranade) [reply data/changelog.txt:29]: These aren't components, weather types for example aren't a visible thing on the map, and neither are terrain transformations.

**Pull request merged**